modin/pandas/test/test_series.py

```diff
@@ -4518,18 +4520,67 @@ def test_struct_general():
     )
 
 
-def test_case_when():
-    # Copied from pandas
-    c_md, c_pd = create_test_series([6, 7, 8, 9], name="c")
-    a_md, a_pd = create_test_series([0, 0, 1, 2])
-    b_md, b_pd = create_test_series([0, 3, 4, 5])
+def _case_when_caselists():
+    def permutations(values):
+        return [
+            p
+            for r in range(1, len(values) + 1)
+            for p in itertools.permutations(values, r)
+        ]
 
-    results = [None, None]
-    for idx, (c, a, b) in enumerate(((c_md, a_md, b_md), (c_pd, a_pd, b_pd))):
-        results[idx] = c.case_when(
-            caselist=[(a.gt(0), a), (b.gt(0), b)]  # condition, replacement
-        )
-    df_equals(*results)
+    conditions = permutations(
+        [
+            [True, False, False, False] * 10,
+            pandas.Series([True, False, False, False] * 10),
+            pandas.Series([True, False, False, False] * 10, index=range(78, -2, -2)),
+            lambda df: df.gt(0),
+        ]
+    )
+    replacements = permutations([[0, 3, 4, 5] * 10, 0, lambda df: 1])
+    caselists = []
+    for c in conditions:
+        for r in replacements:
+            if len(c) == len(r):
+                caselists.append(list(zip(c, r)))
+    return caselists
+
+
+@pytest.mark.parametrize(
+    "base",
+    [
+        pandas.Series(range(40)),
+        pandas.Series([0, 7, 8, 9] * 10, name="c", index=range(0, 80, 2)),
+    ],
+)
+@pytest.mark.parametrize(
+    "caselist",
+    _case_when_caselists(),
+)
+def test_case_when(base, caselist):
+    pd_result = base.case_when(caselist)
+    nparts = NPartitions.get()
+    part_size = MinPartitionSize.get()
+    new_nparts = max(1, min(math.ceil(len(base) / part_size), part_size)) + 1
+    NPartitions.put(new_nparts)
+    MinPartitionSize.put(math.ceil(len(base) / new_nparts))
+    base_repart = pd.Series(base)
+    NPartitions.put(nparts)
+    MinPartitionSize.put(part_size)
```

Comment:
It doesn't work for Series:

```python traceback
modin/core/dataframe/pandas/partitioning/partition_manager.py", line 1043, in <lambda>
E                           index_func = lambda df: df.axes[axis]  # noqa: E731
```

Reply:
?
```python
construct_modin_df_by_scheme(base.to_frame(), ...).squeeze(axis=1)
```

Reply:
It works, but Unidist and HDK tests fail with NotImplementedError - https://github.com/modin-project/modin/actions/runs/8194577509/job/22410966951?pr=6972. On Ray it works, but the test runs ~20% longer.

Reply:
It should be quite easy to add `construct_modin_series_by_scheme` (by analogue). Then there will be no slowdown and the code will be more reliable (there is no way to change variables and not return the previous values).

Reply:
It should be moved to some utility class, because when it's used from test_internals, NotImplementedError is raised for Unidist and HDK.
I don't know why, but the test runs slower. Perhaps, because the constructed from partitions frame does not have the lengths info.

Reply:
I see. Let's use fixtures: https://github.com/modin-project/modin/blob/b1501d8dd838d36842d7eeccd07a2fad9073daf1/modin/conftest.py#L462 and https://github.com/modin-project/modin/blob/b1501d8dd838d36842d7eeccd07a2fad9073daf1/modin/conftest.py#L486 in this case.

Reply:
But, in this case, all the frames will be identically partitioned. The idea is to test cases when the base and the arguments have different partitioning.

Reply:
`NotImplementedError` is raised because the utility function being imported from `modin.test.storage_formats.pandas.test_internals.py` that doesn't support testing nor with Unidist nor with HDK:
https://github.com/modin-project/modin/blob/b1501d8dd838d36842d7eeccd07a2fad9073daf1/modin/test/storage_formats/pandas/test_internals.py#L86-L89
The utility function should work just fine with Unidist if imported from any other place (for HDK we could just skip the test, or make the `construct_by_scheme_function` simply return `pd.DataFrame(data)` since HDK doesn't support modin's partitioning)

Reply:
Added minor optimizations to construct_modin_df_by_scheme(). Now it works for Unidist and for HDK it's not used.